### ISSUE TITLE
multiple feed handlers stall on 1006 error from websockets library

### ISSUE THREAD

**Issue by kyle-downey** [bmoscon/cryptofeed#545]
**Describe the bug**

I ran two processes connecting to Gemini and Coinbase Pro today, starting at around 4pm. At 7:38 p.m. -- simultaneously, two completely separate Python processes -- both experienced ConnectionClosedError, code 1006, and stopped ticking. There was no evidence of reconnection, and my metrics endpoint showed trades & L2 book updates dropped to zero and never recovered. The Gemini feed handler process had a couple errors like this:

```
2021-07-02 19:38:46,677 [MainThread] - asyncio - ERROR - Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason')>
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
```

while Coinbase Pro had more extensive errors, with stack traces:

```
2021-07-02 19:38:59,960 : WARNING : COINBASE.ws.1: received no messages within timeout, restarting connection
2021-07-02 19:39:29,982 [MainThread] - asyncio - ERROR - Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason')>
Traceback (most recent call last):
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 750, in transfer_data
    message = await self.read_message()
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 819, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 895, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 971, in read_frame
    frame = await Frame.read(
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/framing.py", line 55, in read
    data = await reader(2)
  File "/usr/local/Cellar/python@3.8/3.8.10/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/usr/local/Cellar/python@3.8/3.8.10/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
2021-07-02 19:39:30,029 [MainThread] - asyncio - ERROR - Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason')>
Traceback (most recent call last):
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 750, in transfer_data
    message = await self.read_message()
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 819, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 895, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 971, in read_frame
    frame = await Frame.read(
  File "/Users/kdowney/dev/shadow/serenity/venv/lib/python3.8/site-packages/websockets/legacy/framing.py", line 55, in read
    data = await reader(2)
  File "/usr/local/Cellar/python@3.8/3.8.10/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/usr/local/Cellar/python@3.8/3.8.10/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
asyncio.exceptions.CancelledError
```

Below you can find the feedhandler.log as well, which shows that it did try to reconnect at 7:38 p.m., e.g.:

```
2021-07-02 19:38:59,960 : WARNING : COINBASE.ws.1: received no messages within timeout, restarting connection
```

[feedhandler.log](https://github.com/bmoscon/cryptofeed/files/6757687/feedhandler.log)

**To Reproduce**
The code on my side is quite extensive, but the integration with cryptofeed is basically this:

```
        subscription = {TRADES: symbols, L2_BOOK: symbols}
        callbacks = {
            TRADES: TradeCallback(self._on_trade, include_order_type=False),
            BOOK_DELTA: BookUpdateCallback(self._on_book_delta)
        }
        f.add_feed(self._create_feed(subscription=subscription, callbacks=callbacks))
```

where _create_feed() is defined in the subclass, e.g. for Coinbase Pro it is:

```
    def _create_feed(self, subscription: dict, callbacks: dict):
        return Coinbase(subscription=subscription, callbacks=callbacks)
```

Full code here: https://github.com/cloudwall/serenity/blob/master/src/serenity/marketdata/fh/feedhandler.py

See CryptofeedFeedHandler.

**Expected behavior**
I expect automatic reconnects, continuously ticking data and no stack traces.

**Screenshots**
My metrics dashboard shows the gap from when it stalled until when I restarted:

<img width="683" alt="marketdata_dashboard" src="https://user-images.githubusercontent.com/8682922/124340597-4c778780-db84-11eb-9d88-15514e0632cb.png">

**Operating System:**
 - tested on macOS "Big Sur" with Python 3.8

**Cryptofeed Version**
 - cryptofeed 1.9.1, installed from pip

**Comment by kyle-downey** [comment]
attaching the feedhandler.log as well, which shows that it did try to reconnect at 7:38 p.m., e.g.:

```
2021-07-02 19:38:59,960 : WARNING : COINBASE.ws.1: received no messages within timeout, restarting connection
```

[feedhandler.log](https://github.com/bmoscon/cryptofeed/files/6757687/feedhandler.log)

**Comment by kyle-downey** [comment]
Full code here: https://github.com/cloudwall/serenity/blob/master/src/serenity/marketdata/fh/feedhandler.py

See CryptofeedFeedHandler.

**Comment by bmoscon** [comment]
disconnects are impossible to avoid - they can be caused by many things - network issues on your end, network issues at the exchange, a network issue anywhere in the many hops between you and the exchange, planned downtime/maintenance, etc. Some exchanges even mention in their docs that you should expect to be disconnected every 24 hours. Its not really obvious what the cause of this issue is. From your log I see this:

`ConnectionResetError: [Errno 54] Connection reset by peer`

from the first exception. This makes me think it was somewhere upstream of you. It looks like the disconnect then caused some errors in your program:
```
 File "/Users/kdowney/dev/shadow/serenity/src/serenity/marketdata/fh/feedhandler.py", line 164, in _publish_trade_print
    msg = capnp_def.RawFeedTradeMessage.new_message()
AttributeError: module 'feedhandler.capnp' has no attribute 'RawFeedTradeMessage'
2021-06-26 14:37:10,416 : ERROR : COINBASE.ws.1: encountered an exception, reconnecting in 1.0 seconds
```

plus many many more, especially in the cython files

in your issue above you list a CancelledError, which means the asyncio task was killed. I'm not positive, because I don't know your code, but it looks like the disconnect caused errors in your code which caused the cascade of failures preventing it from reconnecting. Once the task is cancelled, it will never restart the connection. It only tries to handle internal exceptions and reconnect. 


As far as the gemini DC issues go. There are two things we can do.

1. Per the gemini docs: `Gemini recommends logging and retaining all heartbeat messages. If your WebSocket connection is unreliable, please contact Gemini support with this log.` I can add support for this, and you can log these and I can take them to gemini support and work with them on it. 
2. They recently modernized part of their connection interface and they now support a method of subscribing to multiple symbols in a single connection. This may alleviate your issue (or may make it worse, can't say unless we try it)

If you'd like to try either of the above, or have a different thing you'd like to try, let me know

**Comment by kyle-downey** [comment]
Thank you so much for coming back on this; it's quite important to me to sort this out so I can switch to cryptofeed from my hand-rolled websockets code. The error you cite is from a number of days ago and was fixed already, so I think that's a red herring. I will think about what you've suggested and go through the logs myself again to see if there's some missing error handling that's causing the task to die. My expectation would be that the 1006 would be caught and handled with a retry inside the cryptofeed code without an exception bubbling up and killing the asyncio task, but maybe I've missed something.

**Comment by bmoscon** [comment]
just going through the log file you attached, it seems like its an endless loop of retries because the user supplied callbacks are crashing

**Comment by kyle-downey** [comment]
> just going through the log file you attached, it seems like its an endless loop of retries because the user supplied callbacks are crashing

hi Bryant, the original posted logfile has traffic from too many sessions while I was debugging. I ran it in my dev installation and all I see is this now, albeit it shows up several times after the initial 1006 error:

```
2021-07-04 19:51:25,911 [MainThread] - asyncio - ERROR - Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason')>
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
```

So far I cannot figure this out. I understand 1006 errors are normal, but the lack of ticking updates after the 1006 is definitely not what I expect given cryptofeed's built-in retries. Any advice on what I can do to help debug this?

**Comment by bmoscon** [comment]
If I'm running the feedhandler and I unplug the network cable and then plug it back in, it will eventually retry and reconnect. What happens if you do that on your end? Does it reconnect or does it die?

**Comment by kyle-downey** [comment]
I disabled WiFi with both the Gemini & Coinbase Pro feed handlers running. I confirmed that before I disabled it I was getting good ticking data, and after I disabled it I got the 1006 errors again and it did not recover: tick rates flatlined to zero and stayed that way, so it did not retry: it dies. This is on MacOS "Big Sur" with Python 3.8.10, asyncio is using EPOLL, and latest cryptofeed from pypi. That is, in a way, good news: it means there's something different about my setup than yours which is blocking retries inside cryptofeed. We just have to figure out what's different ....

**Comment by bmoscon** [comment]
I tried your test on my macOS laptop (also bigsur). I'm running python 3.9.1. I used wifi as well, turned it off, waited for errors, and then turned it back on:


Timestamp: 1625450268.829701 Feed: COINBASE Symbol: BTC-USD Bid: 34599.90 Ask: 34608.59
Timestamp: 1625450268.829701 Cryptofeed Receipt: 1625450268.8667421 Feed: COINBASE Symbol: BTC-USD ID: 192283784 Side: buy Amount: 0.00006 Price: 34599.91
Error in data transfer
Traceback (most recent call last):
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 827, in transfer_data
    message = await self.read_message()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 895, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 971, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 1047, in read_frame
    frame = await Frame.read(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/framing.py", line 105, in read
    data = await reader(2)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/selector_events.py", line 856, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
TimeoutError: [Errno 60] Operation timed out
Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason')>
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
2021-07-04 21:58:13,306 : WARNING : COINBASE.ws.3: encountered connection issue code = 1006 (connection closed abnormally [internal]), no reason - reconnecting in 1.0 seconds...
Traceback (most recent call last):
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 827, in transfer_data
    message = await self.read_message()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 895, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 971, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 1047, in read_frame
    frame = await Frame.read(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/framing.py", line 105, in read
    data = await reader(2)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/selector_events.py", line 856, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
TimeoutError: [Errno 60] Operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 66, in _create_connection
    await self._handler(connection, self.handler)
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 111, in _handler
    async for message in connection.read():
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 219, in read
    async for data in self.conn:
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 439, in __aiter__
    yield await self.recv()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 509, in recv
    await self.ensure_open()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 803, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
Error in data transfer
Traceback (most recent call last):
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 827, in transfer_data
    message = await self.read_message()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 895, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 971, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 1047, in read_frame
    frame = await Frame.read(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/framing.py", line 105, in read
    data = await reader(2)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/selector_events.py", line 856, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
TimeoutError: [Errno 60] Operation timed out
Future exception was never retrieved
future: <Future finished exception=ConnectionClosedError('code = 1006 (connection closed abnormally [internal]), no reason')>
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
2021-07-04 21:58:13,329 : WARNING : COINBASE.ws.2: encountered connection issue code = 1006 (connection closed abnormally [internal]), no reason - reconnecting in 1.0 seconds...
Traceback (most recent call last):
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 827, in transfer_data
    message = await self.read_message()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 895, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 971, in read_data_frame
    frame = await self.read_frame(max_size)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 1047, in read_frame
    frame = await Frame.read(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/framing.py", line 105, in read
    data = await reader(2)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/selector_events.py", line 856, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
TimeoutError: [Errno 60] Operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 66, in _create_connection
    await self._handler(connection, self.handler)
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 111, in _handler
    async for message in connection.read():
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 219, in read
    async for data in self.conn:
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 439, in __aiter__
    yield await self.recv()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 509, in recv
    await self.ensure_open()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/protocol.py", line 803, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
2021-07-04 21:58:14,315 : WARNING : COINBASE.ws.3: encountered connection issue [Errno 8] nodename nor servname provided, or not known - reconnecting in 2.0 seconds...
Traceback (most recent call last):
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 56, in _create_connection
    async with self.conn.connect() as connection:
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/contextlib.py", line 175, in __aenter__
    return await self.gen.__anext__()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 76, in connect
    await self._open()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 204, in _open
    self.conn = await websockets.connect(self.address, **self.ws_kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/client.py", line 535, in __await_impl__
    transport, protocol = await self._create_connection()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1017, in create_connection
    infos = await self._ensure_resolved(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1396, in _ensure_resolved
    return await loop.getaddrinfo(host, port, family=family, type=type,
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/socket.py", line 953, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
2021-07-04 21:58:14,331 : WARNING : COINBASE.ws.2: encountered connection issue [Errno 8] nodename nor servname provided, or not known - reconnecting in 2.0 seconds...
Traceback (most recent call last):
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 56, in _create_connection
    async with self.conn.connect() as connection:
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/contextlib.py", line 175, in __aenter__
    return await self.gen.__anext__()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 76, in connect
    await self._open()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 204, in _open
    self.conn = await websockets.connect(self.address, **self.ws_kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/client.py", line 535, in __await_impl__
    transport, protocol = await self._create_connection()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1017, in create_connection
    infos = await self._ensure_resolved(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1396, in _ensure_resolved
    return await loop.getaddrinfo(host, port, family=family, type=type,
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/socket.py", line 953, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
2021-07-04 21:58:16,331 : WARNING : COINBASE.ws.3: encountered connection issue [Errno 8] nodename nor servname provided, or not known - reconnecting in 4.0 seconds...
Traceback (most recent call last):
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 56, in _create_connection
    async with self.conn.connect() as connection:
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/contextlib.py", line 175, in __aenter__
    return await self.gen.__anext__()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 76, in connect
    await self._open()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 204, in _open
    self.conn = await websockets.connect(self.address, **self.ws_kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/client.py", line 535, in __await_impl__
    transport, protocol = await self._create_connection()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1017, in create_connection
    infos = await self._ensure_resolved(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1396, in _ensure_resolved
    return await loop.getaddrinfo(host, port, family=family, type=type,
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/socket.py", line 953, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
2021-07-04 21:58:16,333 : WARNING : COINBASE.ws.2: encountered connection issue [Errno 8] nodename nor servname provided, or not known - reconnecting in 4.0 seconds...
Traceback (most recent call last):
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection_handler.py", line 56, in _create_connection
    async with self.conn.connect() as connection:
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/contextlib.py", line 175, in __aenter__
    return await self.gen.__anext__()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 76, in connect
    await self._open()
  File "/Users/bryant/source/cryptofeed/cryptofeed/connection.py", line 204, in _open
    self.conn = await websockets.connect(self.address, **self.ws_kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/site-packages/websockets-8.1-py3.9-macosx-10.9-x86_64.egg/websockets/client.py", line 535, in __await_impl__
    transport, protocol = await self._create_connection()
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1017, in create_connection
    infos = await self._ensure_resolved(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 1396, in _ensure_resolved
    return await loop.getaddrinfo(host, port, family=family, type=type,
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/bryant/anaconda3/envs/crypto/lib/python3.9/socket.py", line 953, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
Timestamp: 1625450300.554572 Cryptofeed Receipt: 1625450300.8242528 Feed: COINBASE Symbol: BTC-USD ID: 192283833 Side: sell Amount: 0.01192785 Price: 34582.59
Timestamp: 1625450300.554572 Feed: COINBASE Symbol: BTC-USD Bid: 34582.59 Ask: 34582.60
Timestamp: 1625450301.354762 Cryptofeed Receipt: 1625450301.314871 Feed: COINBASE Symbol: BTC-USD ID: 192283834 Side: buy Amount: 0.00140975 Price: 34580.4



As you can see, no problem restarting the feeds. I was running two coinbase feeds here, one with trades, one with ticker. I'm running the code from master in github though, and of course my versions of libraries may be different. I'm also using uvloop. I tried with websockets 8.1 and 9.1 (the only library likely to make a difference).

**Comment by bmoscon** [comment]
my code was just the demo.py code, except the body of main was just this:

```python

    config = {'log': {'filename': 'demo.log', 'level': 'INFO'}}
    f = FeedHandler(config=config)
    
    f.add_feed(COINBASE, symbols=['BTC-USD'], channels=[TICKER], callbacks={TICKER: TickerCallback(ticker)})
    f.add_feed(Coinbase(symbols=['BTC-USD'], channels=[TRADES], callbacks={TRADES: TradeCallback(trade)}))

    f.run()

```

**Comment by kyle-downey** [comment]
OK, I will try Python 3.9.x and uvloop independently to see if either makes any difference, and I will see if I can reproduce using my library stack and your demo.py code.

**Comment by kyle-downey** [comment]
So, some results:

1. switching to Python 3.9.x made no difference
2. switching to uvloop made no difference
3. switching to the master branch made no difference

So I started setting breakpoints to see where it's hanging, and I think I found the issue. WSAsyncConn._open() is the culprit, specifically this line, which attempts to connect without a timeout on the event loop:

```
            self.conn = await websockets.connect(self.address, **self.ws_kwargs)
```

At least on my Mac what seems to be happening is this call never returns, and so the whole retry loop freezes; that's why it doesn't reconnect. If I just change this one line to introduce a five second connection timeout:

```
            self.conn = await asyncio.wait_for(websockets.connect(self.address, **self.ws_kwargs), 5)
```

after 5 seconds the connect() operation times out, and the retry loop attempts again as expected. I am able to enable and disable WiFi and stop and start ticking data, e.g. now I see this with DEBUG logging:

```
2021-07-05 09:37:29,871 : DEBUG : COINBASE.ws.1: connecting to wss://ws-feed.pro.coinbase.com
2021-07-05 09:37:30,060 : WARNING : COINBASE.ws.1: received no messages within timeout, restarting connection
2021-07-05 09:37:30,136 : INFO : COINBASE.ws.1: closed connection 'WebSocketClientProtocol'
2021-07-05 09:37:30,143 : DEBUG : COINBASE.ws.1: connecting to wss://ws-feed.pro.coinbase.com
```

Now what I cannot tell you is *why* this only happens to me and works fine for you, but I do think await websockets.connect() without a timeout is worth fixing regardless. Let me know what you think.
